### PR TITLE
[fix][sft] Locate the assistant response boundary instead of subtracting prompt length

### PR DESCRIPTION
**Pull request by bharatmekala**
## Summary

`_tokenize_chat_last_assistant` derives the supervised window by tokenizing the conversation twice and subtracting encoded lengths:

```python
num_actions = len(full_input_ids) - len(full_prompt_ids)
```

That is only correct when the generation-prompt encoding is a token *prefix* of the full-conversation encoding. It isn't when the tokenizer merges the prompt's trailing `\n` with the response's leading whitespace into one `\n\n` token — the exact merge `_find_generation_prompt_boundary` was added to handle for `ALL_ASSISTANT_MESSAGES` in #1611. The last-assistant path never got that treatment, so on those rows the prompt is one token longer than the corresponding span of the full sequence, `num_actions` comes out one short, and the first token the assistant actually generated silently trains with weight 0.

On `Qwen/Qwen2.5-0.5B-Instruct` with `[{"role": "user", "content": "Hi"}, {"role": "assistant", "content": "\nHello world"}]` (the leading-newline shape TULU3-style data produces):

| | |
|---|---|
| prompt, `add_generation_prompt=True` | 30 tokens |
| full conversation | 34 tokens |
| supervised before | 4 → `'Hello world<\|im_end\|>\n'` |
| true response | 5 → `'\n\nHello world<\|im_end\|>\n'` |

The merged `\n\n` (id 271) belongs to the assistant but is never supervised. With `"Hello world"` the prompt *is* a prefix and the subtraction is right, so only the merge case is affected.

So instead of subtracting lengths, locate where the prompt actually ends inside the full sequence:

```python
prompt_boundary = _find_generation_prompt_boundary(full_input_ids, full_prompt_ids)
num_actions = len(full_input_ids) - prompt_boundary
```

The merged boundary token gets weight 1, i.e. attributed to the assistant's generation — the same choice the all-assistant path makes, so both `train_on_what` modes now supervise the same tokens for a given conversation.

Two details worth calling out:

- The helper raises when neither the exact-prefix nor the merge pattern matches. That is caught, logged, and falls back to the previous length-based boundary, so an unusual chat template degrades to today's behavior instead of aborting a run partway through a dataset.
- A guard returns `None` when truncation caps both encodings at `max_length` and leaves no response tokens. That keeps truncation out of the boundary lookup (it isn't a merge, and would otherwise warn on every over-length row) and makes the old `num_actions <= 0` check unreachable, so it is removed.

`ALL_ASSISTANT_MESSAGES` is untouched.

## Test plan

- [x] `uv run --isolated --extra dev --extra fsdp pytest tests/train/test_sft_tokenization.py -v` — 38 passed
- [x] `uv run --isolated --extra dev --extra fsdp pytest tests/train/generators/test_utils.py -v` — 58 passed
- [x] `ruff` / `black` / secret check via pre-commit

New CPU tests, all on the smallest Qwen2.5 instruct model:

- `test_last_assistant_leading_newline_merge` — supervised window is exactly `[271, 9707, 1879, 151645, 198]` (`'\n\nHello world<|im_end|>\n'`), including the merged boundary token. Fails on `main` with `num_actions == 4`.
- `test_last_assistant_without_newline_merge` — control: with prefix-matching encodings `num_actions` still equals `len(full) - len(prompt)`, so the common path is unmoved.
- `test_chat_truncation_prompt_fills_budget` — `max_length` at the prompt length returns `None`.

---

> [!NOTE]
> **Medium Risk**
> Changes which tokens receive loss in last-assistant SFT for merge-prone chat templates; behavior is intentional but will alter gradients on affected rows compared to main.
> 
> **Overview**
> Fixes **last-assistant** SFT tokenization so the supervised window matches the real assistant span when the chat template **merges** the generation prompt’s trailing newline with leading response whitespace (e.g. Qwen2.5 + `\nHello world`).
> 
> **`_tokenize_chat_last_assistant`** no longer uses `len(full) - len(prompt)`; it calls **`_find_generation_prompt_boundary`** (same helper as the all-assistant path) so the merged boundary token is included in `num_actions` / `loss_mask`. If the boundary cannot be found, it **logs and falls back** to the old length-based split. Truncation that leaves **no response tokens** now returns **`None`** via an explicit length check instead of `num_actions <= 0`.
> 
> Adds **Qwen2.5** CPU tests for the merge case, the prefix-matching control, and truncation when `max_length` equals prompt length.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0befcdffeb062f5c3412357a072f3a84f1fa527c. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/train/test_sft_tokenization.py [hunk 1]
```diff
@@ -126,6 +133,66 @@ def test_chat_truncation(tokenizer):
...
+    """Returns None when truncation removes the complete assistant response."""
+    messages = [
+        {"role": "user", "content": "Tell me a story."},
+        {"role": "assistant", "content": "Once upon a time."},
+    ]
+    prompt_ids = qwen25_tokenizer.apply_chat_template(
+        messages[:-1],
+        add_generation_prompt=True,
+        tokenize=True,
+    )["input_ids"]
```
[the remark this answers is not in the document]
Comment (bharatmekala) [reply tests/train/test_sft_tokenization.py:146]: Thanks, but I don't think this holds for the transformers versions this repo pins. `pyproject.toml` requires `transformers>=5.6.1,<=5.8.0`, and in transformers 5.x `PreTrainedTokenizerBase.apply_chat_template` has `return_dict=True` as its default, so `tokenize=True` returns a `BatchEncoding`, not a bare list:

```
$ python -c "import inspect, transformers; print(transformers.__version__, inspect.signature(transformers.PreTrainedTokenizerBase.apply_chat_template).parameters['return_dict'].default)"
5.6.1 True
5.8.0 True
```

The `["input_ids"]` indexing is therefore required, and the tests pass as written:

```
$ uv run --isolated --extra dev --extra fsdp pytest tests/train/test_sft_tokenization.py -k "newline or budget" -q
3 passed, 35 deselected
```

Applying the suggestion would leave `prompt_ids` as a `BatchEncoding`, making `len(prompt_ids)` 2 (the number of keys) rather than the token count, which silently breaks the `max_length` this test depends on. Leaving the code as is.

File: tests/train/test_sft_tokenization.py [hunk 2]
```diff
@@ -126,6 +133,66 @@ def test_chat_truncation(tokenizer):
...
+    """A prefix-matching prompt keeps the existing response boundary."""
+    messages = [
+        {"role": "user", "content": "Hi"},
+        {"role": "assistant", "content": "Hello world"},
+    ]
+
+    prompt_ids = qwen25_tokenizer.apply_chat_template(messages[:-1], add_generation_prompt=True, tokenize=True)[
+        "input_ids"
+    ]
+    full_ids = qwen25_tokenizer.apply_chat_template(messages, add_generation_prompt=False, tokenize=True)["input_ids"]
```
[the remark this answers is not in the document]
Comment (bharatmekala) [reply tests/train/test_sft_tokenization.py:183]: Same as the thread above: with `transformers>=5.6.1,<=5.8.0` (the range this repo pins), `apply_chat_template(..., tokenize=True)` defaults to `return_dict=True` and returns a `BatchEncoding`, so `["input_ids"]` is needed on both calls. Verified on both ends of the pinned range, and `tests/train/test_sft_tokenization.py -k "newline or budget"` passes as written. Removing the indexing would make `full_ids[: len(prompt_ids)]` compare dict keys instead of tokens.
